### PR TITLE
Update deprecated openAI mode in integration test and fix flaky tests

### LINKED ISSUE #1848: [BUG] replace deprecated model text-davinci-edit-001 and text-davinci-003 in integration test
**What is the bug?**

`text-davinci-001` has been deprecated according to https://platform.openai.com/docs/deprecations
we need to update it to recommended replacement in integration test [RestMLRemoteInferenceIT](https://github.com/opensearch-project/ml-commons/blob/main/plugin/src/test/java/org/opensearch/ml/rest/RestMLRemoteInferenceIT.java)
```
"error": {
?          "message": "The model `text-davinci-edit-001` has been deprecated, learn more here: https://platform.openai.com/docs/deprecations",
?          "type": "invalid_request_error",
?          "param": null,
?          "code": "model_not_found"
?      }
```

```
?  org.opensearch.OpenSearchStatusException: Error from remote service: {
?      "error": {
?          "message": "The model `text-davinci-003` has been deprecated, learn more here: https://platform.openai.com/docs/deprecations",
?          "type": "invalid_request_error",
?          "param": null,
?          "code": "model_not_found"
?      }
?  }
```

**How can one reproduce the bug?**

**What is the expected behavior?**

**What is your host/environment?**
 - Plugins

**Do you have any screenshots?**

**Do you have any additional context?**
Add any other context about the problem.

### PR DESCRIPTION
**Pull request by Hailong-am**
### Description
This PR includes changes for below items
1. pick from #1734
2. part of #1733 
3. replace deprecated openAI mode in integration test
4. update default admin password for integration test with docker run
5. fix getConversations when index not exists missing return
 
### Issues Resolved
#1848 and https://github.com/opensearch-project/ml-commons/issues/1849
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ylwu-amzn** [comment]
Thanks Hailong, I see the CI on linux docker also failed. I know that's not from your change. If you have time, can you help take a look?

**Comment by Hailong-am** [comment]
> Thanks Hailong, I see the CI on linux docker also failed. I know that's not from your change. If you have time, can you help take a look?

sure, will take a look

File: .github/workflows/CI-workflow.yml [hunk 1]
```diff
@@ -145,20 +145,20 @@ jobs:
         if: env.imagePresent == 'true'
         run: |
           cd ..
-          docker run -p 9200:9200 -d -p 9600:9600 -e "discovery.type=single-node" opensearch-ml:test
+          docker run -p 9200:9200 -d -p 9600:9600 -e "discovery.type=single-node" -e OPENSEARCH_INITIAL_ADMIN_PASSWORD=myStrongPassword123! opensearch-ml:test
```
Comment (Zhangxunmt) [inline .github/workflows/CI-workflow.yml:148]: Why not just use the previous default username and password admin:admin for BWC?

**Review by Zhangxunmt — COMMENTED** [review]
Looks like ITs are still failing on dockers. Can you try using the old password admin:admin?

File: .github/workflows/CI-workflow.yml (hunk 1 above)
Comment (Hailong-am) [reply .github/workflows/CI-workflow.yml:148]: admin is too weak, need have a strong one.

**Comment by Hailong-am** [comment]
> Looks like ITs are still failing on dockers. Can you try using the old password admin:admin?

By looking at the logs, CI still using old workflow files. It's wired.

```shell
Run security=`curl -XGET https://localhost:9200/_cat/plugins?v -u admin:admin --insecure |grep opensearch-security|wc -l`
  security=`curl -XGET https://localhost:9[2](https://github.com/opensearch-project/ml-commons/actions/runs/7477731642/job/20364986522?pr=1846#step:7:2)00/_cat/plugins?v -u admin:admin --insecure |grep opensearch-security|wc -l`
  export OPENAI_KEY=$(aws secretsmanager get-secret-value --secret-id github_openai_key --query SecretString --output text)
  export COHERE_KEY=$(aws secretsmanager get-secret-value --secret-id github_cohere_key --query SecretString --output text)
```

**Comment by Hailong-am** [comment]
> > Looks like ITs are still failing on dockers. Can you try using the old password admin:admin?
> 
> By looking at the logs, CI still using old workflow files. It's wired.
> 
> ```shell
> Run security=`curl -XGET https://localhost:9200/_cat/plugins?v -u admin:admin --insecure |grep opensearch-security|wc -l`
>   security=`curl -XGET https://localhost:9[2](https://github.com/opensearch-project/ml-commons/actions/runs/7477731642/job/20364986522?pr=1846#step:7:2)00/_cat/plugins?v -u admin:admin --insecure |grep opensearch-security|wc -l`
>   export OPENAI_KEY=$(aws secretsmanager get-secret-value --secret-id github_openai_key --query SecretString --output text)
>   export COHERE_KEY=$(aws secretsmanager get-secret-value --secret-id github_cohere_key --query SecretString --output text)
> ```

As the doc said https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#controlling-changes-from-forks-to-workflows-in-public-repositories, it will always run with base branch for security consideration.

**Comment by Hailong-am** [comment]
> > > Looks like ITs are still failing on dockers. Can you try using the old password admin:admin?
> > 
> > 
> > By looking at the logs, CI still using old workflow files. It's wired.
> > ```shell
> > Run security=`curl -XGET https://localhost:9200/_cat/plugins?v -u admin:admin --insecure |grep opensearch-security|wc -l`
> >   security=`curl -XGET https://localhost:9[2](https://github.com/opensearch-project/ml-commons/actions/runs/7477731642/job/20364986522?pr=1846#step:7:2)00/_cat/plugins?v -u admin:admin --insecure |grep opensearch-security|wc -l`
> >   export OPENAI_KEY=$(aws secretsmanager get-secret-value --secret-id github_openai_key --query SecretString --output text)
> >   export COHERE_KEY=$(aws secretsmanager get-secret-value --secret-id github_cohere_key --query SecretString --output text)
> > ```
> 
> As the doc said https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#controlling-changes-from-forks-to-workflows-in-public-repositories, it will always run with base branch for security consideration.

@ylwu-amzn  @Zhangxunmt I have run the docker integration test on forked personal repo, all test has passed https://github.com/Hailong-am/ml-commons/actions/runs/7485221729/job/20373533131?pr=3.

**Comment by dhrubo-os** [comment]
Thanks for raising this PR.

Can we also change here:

1. https://github.com/opensearch-project/ml-commons/blob/main/plugin/src/test/java/org/opensearch/ml/utils/TestHelper.java#L230

2. https://github.com/opensearch-project/ml-commons/blob/main/docs/remote_inference_blueprints/open_ai_connector_completion_blueprint.md

**Comment by dhrubo-os** [comment]
Don't we need to change here: https://github.com/opensearch-project/ml-commons/blob/deb51f6163d884bff5db3082ab661bc53b426677/plugin/src/test/java/org/opensearch/ml/rest/RestMLRemoteInferenceIT.java#L268C35-L268C56


according to the [deprecation doc](https://platform.openai.com/docs/deprecations), `text-davinci-edit-001` should be replaced by `gpt-4` and `/v1/edits` needs to be replaced by `	/v1/chat/completions`

If you reflect those changes as well, let's update the corresponding blue print docs as well.

**Comment by Hailong-am** [comment]
> according to the [deprecation doc](https://platform.openai.com/docs/deprecations), `text-davinci-edit-001` should be replaced by `gpt-4` and `/v1/edits` needs to be replaced by ` /v1/chat/completions`

1. whether our OPENAI key have access to gpt-4?
2. the endpoint replaced by `/v1/chat/completions`, it duplicate with [testOpenAIChatCompletionModel](https://github.com/opensearch-project/ml-commons/blob/main/plugin/src/test/java/org/opensearch/ml/rest/RestMLRemoteInferenceIT.java#L194), so in this PR i marked testOpenAIEditsModel as ignored

**Comment by Hailong-am** [comment]
> 1. https://github.com/opensearch-project/ml-commons/blob/main/plugin/src/test/java/org/opensearch/ml/utils/TestHelper.java#L230

This is a unit test and don't use the connector to do predict, but anyway let's have it updated

> 2. https://github.com/opensearch-project/ml-commons/blob/main/docs/remote_inference_blueprints/open_ai_connector_completion_blueprint.md

updated

File: plugin/src/test/java/org/opensearch/ml/rest/RestMLRemoteInferenceIT.java [hunk 2]
```diff
@@ -251,6 +252,7 @@ public void testOpenAIChatCompletionModel() throws IOException, InterruptedExcep
         assertNotNull(responseMap);
     }
 
+    @Ignore("text-davinci-edit-001 been deprecated on 2024-01-04 and replaced by /v1/chat/completions")
```
Comment (dhrubo-os) [inline plugin/src/test/java/org/opensearch/ml/rest/RestMLRemoteInferenceIT.java:255]: why do we need to ignore this test? I updated here : https://github.com/opensearch-project/ml-commons/pull/1835/commits/5021cfd77c5f265815dafde17d94835ee5498c0b#diff-1551321199e798aacb50228526ec9e02603234147ac41c7ef228ff2e2954e323


But you can add this in your code as you have other changes also.

**Comment by dhrubo-os** [comment]
> > according to the [deprecation doc](https://platform.openai.com/docs/deprecations), `text-davinci-edit-001` should be replaced by `gpt-4` and `/v1/edits` needs to be replaced by ` /v1/chat/completions`
> 
>     1. whether our OPENAI key have access to gpt-4?
> 
>     2. the endpoint replaced by `/v1/chat/completions`, it duplicate with [testOpenAIChatCompletionModel](https://github.com/opensearch-project/ml-commons/blob/main/plugin/src/test/java/org/opensearch/ml/rest/RestMLRemoteInferenceIT.java#L194), so in this PR i marked testOpenAIEditsModel as ignored

I think Our api key has access to all the models.

File: plugin/src/test/java/org/opensearch/ml/rest/RestMLRemoteInferenceIT.java (hunk 2 above)
Comment (Hailong-am) [reply plugin/src/test/java/org/opensearch/ml/rest/RestMLRemoteInferenceIT.java:255]: I think this is duplicate with `testOpenAIChatCompletionModel` which use gpt-3.5 and the endpoint is same, do we need to have two test cases test same functionality?

Comment (Hailong-am) [reply plugin/src/test/java/org/opensearch/ml/rest/RestMLRemoteInferenceIT.java]: update to gpt-4. If we think keep just one is enough, we can remove it in future PR.

**Review by Zhangxunmt — DISMISSED** [review]

File: .github/workflows/CI-workflow.yml (hunk 1 above)
Comment (ylwu-amzn) [reply .github/workflows/CI-workflow.yml:148]: Let's avoid using hard coded password in code

Comment (Hailong-am) [reply .github/workflows/CI-workflow.yml]: done by replaced with generated one

**Review by dhrubo-os — APPROVED** [review]

File: .github/workflows/CI-workflow.yml (hunk 1 above)
Comment (Hailong-am) [reply .github/workflows/CI-workflow.yml]: docker run success in my fork repo https://github.com/Hailong-am/ml-commons/actions/runs/7497254147/job/20410737212

**Review by ylwu-amzn — APPROVED** [review]

**Pull request merged**